node/actors/bft/src/testonly/twins/scenario.rs

```diff
@@ -0,0 +1,182 @@
+use std::collections::BTreeSet;
+
+use rand::{seq::SliceRandom, Rng};
+
+use super::{partitions, HasKey, Partitioning, Twin};
+
+/// A cluster holds all the nodes in the simulation, some of which are twins.
+pub struct Cluster<T> {
+    /// All nodes, twins and originals.
+    nodes: Vec<T>,
+    /// The ideal size of the committee, ie. the `n` in the `n = 5 * f + 1`.
+    num_replicas: usize,
+}
+
+impl<T> Cluster<T>
+where
+    T: Twin,
+{
+    /// Take a number of non-twin replicas and produce a cluster with a desired number of twins.
+    pub fn new(replicas: Vec<T>, num_twins: usize) -> Self {
+        // Check that the replicas aren't twins.
+        assert_eq!(
+            unique_key_count(&replicas),
+            replicas.len(),
+            "replicas must have unique keys"
+        );
+
+        let twins = replicas
+            .iter()
+            .take(num_twins)
+            .map(|r| r.to_twin())
+            .collect::<Vec<_>>();
+
+        let num_replicas = replicas.len();
+
+        let mut nodes = replicas;
+        nodes.extend(twins);
+
+        Self {
+            nodes,
+            num_replicas,
+        }
+    }
+
+    /// The number of nodes in the cluster, both replicas and twins.
+    pub fn num_nodes(&self) -> usize {
+        self.nodes.len()
+    }
+
+    /// The number of original replicas, ie. not twins.
+    ///
+    /// This is the `n` in the `n = 5 * f + 1` equation.
+    pub fn num_replicas(&self) -> usize {
+        self.num_replicas
+    }
+
+    /// The number of twins in the cluster.
+    pub fn num_twins(&self) -> usize {
+        self.num_nodes() - self.num_replicas()
+    }
+
+    /// The number of faulty nodes the committee can tolerate for the `n = 5 * f + 1` relation to hold.
+    pub fn num_faulty(&self) -> usize {
+        (self.num_replicas - 1) / 5
+    }
+
+    /// The number of signatures required for a quorum, ie. `4 * f + 1`
+    pub fn quorum_size(&self) -> usize {
+        self.num_faulty() * 4 + 1
```

Comment:
in general quorum is n-f, not 4f+1. IMO it would be good to stick to that notation, even if your code only supports n=5f+1.

Reply:
In our literature I have always seen 4*f+1, so I thought for anyone reading this, it's more helpful as a distinction. 

For example if we take a look at https://www.notion.so/matterlabs/Two-step-Fast-HotStuff-487def0222f14bb780648357380691e8 I count 14 instances of "4f+1" and exactly 1 "n-f"

But I agree, `n-f` is more general and we can change the equation to be more general, no problem.

Reply:
And if `f = 0` it will actually be `n` which is better than 1 👍

Reply:
Hm, actually if we consider n=5f+1 and take n=7, then f=1, and the 4f+1 the replicas would be waiting for is 5, yet n-f is 6.

Reply:
For a normal consensus it would be `quorum_size = (size + maxFaulty) / 2 + 1`

Reply:
Okay, so the spec says 4f+1, but it's just a theoretical value that we don't actually use. Like most "normal" consensus that use n*2/3+1, we use the equivalent of n*4/5+1, at least the new tests I added pass with this criteria.